############################################################
# Select fields in a proper order 
###########################################################
.mode tabs
.headers on

SELECT 
        variant_id,
  	Gene,
	Ref || '>'||gts AS Variant,
        Chrom,
        Start AS Coordinate,
        Ref,
        Alt,
	VType AS Type,
        VSubType AS Sub_Type,
        Call_rate,
	CASE SUBSTR(gts,1,INSTR(gts,'/')-1)=SUBSTR(gts,INSTR(gts,'/')+1) 
          WHEN 1 THEN 'hom' 
          WHEN 0 THEN 'het' 
        END AS Genotype,
	CASE Is_exonic
          WHEN 1 THEN 'yes' 
          WHEN 0 THEN 'no' 
        END AS Exonic,
        Exon,
        Intron,
	Num_hom_ref,
        Num_het,
        Num_hom_alt,
        Num_unknown,
        Filter,
        Quality,
        gt_quals AS GT_Quality,
	gt_alt_freqs AS Alt_Variant_Freq,
	Depth AS Read_Depth,
	gt_alt_depths AS Alt_Read_Depth,
        gt_ref_depths || ',' || gt_alt_depths AS Allelic_Depths,
        ENTREZ,
	Ensembl_geneid,
        Gene_description,
        OMIM,
        ORPHA,
        NM AS Refseq_Transcript,
        Ensembl_transcriptid AS Ensembl_Transcript,
        Impact_so AS Consequence,
        Impact_severity,
	Cdna_position,
        Cds_position,
        Cds_strand,
	Codon_change AS Codons,
        AA_change AS Amino_Acid_Change,
	Num_refseq_impacts,
	Refseq_impacts,
  	Num_ensembl_impacts INT,
  	Ensembl_impacts,
	CASE Is_canonical
          WHEN 1 THEN 'yes' 
          WHEN 0 THEN 'no' 
        END AS Canonical,
	CASE Is_splicing
          WHEN 1 THEN 'yes' 
          WHEN 0 THEN 'no' 
        END AS Splicing,
	CASE Is_lof
          WHEN 1 THEN 'yes' 
          WHEN 0 THEN 'no' 
        END AS Loss_of_function,
	Cadd_phred,
 	Cadd_raw,
  	Cadd_raw_rankscore,	
	Sift_pred,
        Sift_score,
	Polyphen_pred,
 	Polyphen_score,
  	Ensp,
	Swissprot,
  	Appris,
	Ccds,  
	Hgvsc,
  	Hgvsp,
        RS AS dbSNP_ID,
	Integrated_fitcons_score,
 	Ancestral_allele,
	Gnomad_exome_ac,
	Gnomad_exome_an,
  	Gnomad_exome_af,
  	Gnomad_exome_nhomalt,
  	Gnomad_exome_af_popmax,
  	Gnomad_exome_af_afr,
  	Gnomad_exome_af_amr,
  	Gnomad_exome_af_asj,
  	Gnomad_exome_af_eas,
  	Gnomad_exome_af_sas,
  	Gnomad_exome_af_fin,
  	Gnomad_exome_af_nfe,
  	Gnomad_exome_af_oth,
  	Gnomad_genome_ac,
  	Gnomad_genome_an,
  	Gnomad_genome_af,
  	Gnomad_genome_nhomalt,
 	Gnomad_genome_af_popmax,
  	Gnomad_genome_af_afr,
  	Gnomad_genome_af_amr,
  	Gnomad_genome_af_asj,
  	Gnomad_genome_af_eas,
  	Gnomad_genome_af_sas,
  	Gnomad_genome_af_fin,
  	Gnomad_genome_af_nfe,
  	Gnomad_genome_af_oth,
  	hg19_chr,
  	hg19_pos,
  	Clinvar_disease_name,
  	Clinvar_sig,
  	Codon_degeneracy,
  	Codonpos,
 	Common_pathogenic,
        Dann_rankscore,
        Dann_score,
        Eigen_pc_phred,
        Eigen_pc_raw,
        Eigen_pc_raw_rankscore,
        Eigen_phred,
        Eigen_raw,
        Eigen_coding_or_noncoding,
        Mutationassessor_uniprotid,
        Mutationassessor_pred,
        Mutationassessor_score,
        Mutationassessor_score_rankscore,
        Mutationassessor_variant,
        Mutationtaster_aae,
        Mutationtaster_converted_rankscore,
        Mutationtaster_model,
        Mutationtaster_pred,
        Mutationtaster_score,
        Revel_rankscore,
        Revel_score,
        Vest3_rankscore,
        Vest3_score,
        Fathmm_mkl_coding_group,
        Fathmm_mkl_coding_pred,
        Fathmm_mkl_coding_rankscore,
        Fathmm_mkl_coding_score,
        Integrated_confidence_value,
        Refcodon,
        Allele,
        Feature_type,
        Biotype,        
        Existing_variation,
        Allele_num,
        Distance,
        Strand,
        Flags,
        Pick,
        Variant_class,
        Symbol_source,
        Hgnc_id,
        Tsl,
        Source,
        Given_ref,
        Used_ref,
        Domains,
        Hgvsg
    FROM  variant_fields_gts_genelevel_nm;
